controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -158,47 +186,25 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, kcp *cont
 	}
 	logger = logger.WithValues("cluster", cluster.Name)
 
-	labelSelector := generateKubeadmControlPlaneSelector(cluster.Name)
-	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
-	if err != nil {
-		// Since we are building up the LabelSelector above, this should not fail
-		logger.Error(err, "failed to parse label selector")
-		return ctrl.Result{Requeue: true}
-	}
-	// Copy label selector to its status counterpart in string format.
-	// This is necessary for CRDs including scale subresources.
-	kcp.Status.Selector = selector.String()
-
-	selectorMap, err := metav1.LabelSelectorAsMap(labelSelector)
-	if err != nil {
-		// Since we are building up the LabelSelector above, this should not fail
-		logger.Error(err, "failed to convert label selector to a map")
-		return ctrl.Result{Requeue: true}
-	}
-
-	// Get all Machines linked to this KubeadmControlPlane.
-	allMachines := &clusterv1.MachineList{}
-	err = r.Client.List(
-		context.Background(),
-		allMachines,
-		client.InNamespace(kcp.Namespace),
-		client.MatchingLabels(selectorMap),
+	// TODO: handle proper adoption of Machines
+	ownedMachines, err := getOwnedMachines(
+		ctx,
+		r.Client,
+		kcp,
+		types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Name},
 	)
 	if err != nil {
-		logger.Error(err, "failed to list machines")
+		logger.Error(err, "failed to get list of owned machines")
 		return ctrl.Result{Requeue: true}
 	}
 
-	// Filter out any machines that are not owned by this controller
-	// TODO: handle proper adoption of Machines
-	var ownedMachines []*clusterv1.Machine
-	for i := range allMachines.Items {
-		m := allMachines.Items[i]
-		controllerRef := metav1.GetControllerOf(&m)
-		if controllerRef != nil && controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == kcp.Name {
-			ownedMachines = append(ownedMachines, &m)
+	// Always attempt to update status
+	defer func() {
+		if updateErr := r.updateStatus(ctx, kcp, cluster); updateErr != nil {
+			logger.Error(updateErr, "Failed to update status")
+			res.Requeue = true
```

Comment:
Note (for here & elsewhere): controller-runtime records slightly different metrics for "returned errors" vs "requeue = true". Do you have a preference? I would think we'd want it to put it in the errors category, wouldn't we (which would require returning both the result & an error)?

Reply:
I took this approach due to earlier discussions where we wanted to standardize logging (https://github.com/kubernetes-sigs/cluster-api/issues/1840). I'm fine switching things back up to return the error to controller-runtime, but then we get back into different types of logging based on where the error was handled.

Reply:
Right, if we return an error, we also need to log here so we don't lose the location (as the controller-runtime log message on error is useless, location-wise). We should probably have a discussion with the controller-runtime folks about it.

Reply:
Okay, updated to log and return errors instead of manipulating the requeue value on the result.